pkg/services/ngalert/state/cache.go

```diff
@@ -15,28 +15,38 @@ import (
 	ngModels "github.com/grafana/grafana/pkg/services/ngalert/models"
 )
 
+type ruleStates struct {
+	states map[string]*State
+}
+
 type cache struct {
-	states      map[int64]map[string]map[string]*State // orgID > alertRuleUID > stateID > state
-	mtxStates   sync.RWMutex
-	log         log.Logger
-	metrics     *metrics.State
-	externalURL *url.URL
+	states    map[int64]map[string]*ruleStates // orgID > alertRuleUID > stateID > state
+	mtxStates sync.RWMutex
 }
 
-func newCache(logger log.Logger, metrics *metrics.State, externalURL *url.URL) *cache {
+func newCache() *cache {
 	return &cache{
-		states:      make(map[int64]map[string]map[string]*State),
-		log:         logger,
-		metrics:     metrics,
-		externalURL: externalURL,
+		states: make(map[int64]map[string]*ruleStates),
 	}
 }
 
-func (c *cache) getOrCreate(ctx context.Context, alertRule *ngModels.AlertRule, result eval.Result, extraLabels data.Labels) *State {
+func (c *cache) getOrCreate(ctx context.Context, log log.Logger, alertRule *ngModels.AlertRule, result eval.Result, extraLabels data.Labels, externalURL *url.URL) *State {
 	c.mtxStates.Lock()
 	defer c.mtxStates.Unlock()
+	var ok bool
+	if _, ok = c.states[alertRule.OrgID]; !ok {
```

Comment:
You can simplify this to do fewer lookups:
```go
var orgRules map[string]*ruleStates
var ok bool // necessary to avoid shadowing "orgRules"
if orgRules, ok = c.states[alertRule.OrgID]; !ok {
	orgRules =  make(map[string]*ruleStates)
 	c.states[alertRule.OrgID] = orgRules
}

var states *ruleStates
if states, ok = orgRules[alertRule.UID]; !ok {
 	states = &ruleStates{states: make(map[string]*State)}
 	orgRules[alertRule.UID] = states
}

return states.getOrCreate(ctx, log, alertRule, result, extraLabels, externalURL)
```